components/brave_news/browser/topics_fetcher.h

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_COMPONENTS_BRAVE_NEWS_BROWSER_TOPICS_FETCHER_H_
+#define BRAVE_COMPONENTS_BRAVE_NEWS_BROWSER_TOPICS_FETCHER_H_
+
+#include <string>
+#include <vector>
+
+#include "base/functional/callback_forward.h"
+#include "brave/components/api_request_helper/api_request_helper.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+
+namespace brave_news {
+
+constexpr char kTopicsEndpoint[] = "/news-topic-clustering/topics";
+constexpr char kTopicArticlesEndpoint[] = "/news-topic-clustering/topics_news";
+
+struct TopicArticle {
+  TopicArticle();
+  TopicArticle(const TopicArticle&);
+  TopicArticle(TopicArticle&&);
+  TopicArticle& operator=(const TopicArticle&);
+  ~TopicArticle();
+
+  std::string title;
+  std::string description;
+  GURL url;
+  GURL img;
+  std::string publisher_name;
+  base::Time publish_time;
+  double score;
+  std::string category;
+  std::string origin;
+};
+
+struct Topic {
```

Comment:
I'm on the fence, but what if we remove `Topic` and `TopicArticle`, which are a intermediate types. The only use for these types are for `GenerateTopicBlock` and we could pass the raw data from the idl directly to the function. Current implementation is more like complying 'SOLID' but I usually try to reduce the intermediate types, so still not sure what's the best approach hear.

Reply:
Yeah, I was umming and aahing about that a bit too :laughing: In the end I decided to leave this in because (eventually) we're going to want to send these structs to the front end, so we'll need the conversion logic anyway.

Reply:
Yeah, fair enough! I'm okay with the current impl.

Reply:
Have you considered maybe having this intermediate type codified in an IDL though, as it will already provide ways to serialise it to `base::Value`, and reduce the code maintenance.

Reply:
To be honest, I would prefer we avoided having this intermediate type, **if possible**, and once it becomes a requirement, then we could have such types as part of a new IDL. I understand that in the future this data will have to be massaged to be sent to the front end, but I think it might be better to leave that to be done at its own time, when it is clearer what exactly such a type would look like.

Reply:
> we're going to want to send these structs to the front end

I think @cdesouza-chromium is right. I also assume this should be solvable somehow via an already existing IDL codegen.

But if not, and we want to send it to the front end via mojo, then it would be the right thing to implement these as mojo structs instead of bare c++ ones.

Reply:
I've removed these for now - when we introduce the mojom structs we'll add these conversions.